File: src/python/xerxes/channels/adapters/telegram.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Telegram channel adapter.

Wraps the Telegram Bot API: parses inbound webhook payloads into
``ChannelMessage`` objects and sends replies via ``sendMessage``. The
adapter intentionally stays unaware of authentication — webhook
authenticity and sender allowlisting happen in
``channels.telegram_gateway``.
"""

from __future__ import annotations

import typing as tp

from .._helpers import WebhookChannel, http_post, parse_json_body
from ..types import ChannelMessage, MessageDirection


class TelegramChannel(WebhookChannel):
    """Telegram Bot API adapter."""

    name = "telegram"

    def __init__(self, token: str, *, http_client: tp.Any = None, accept_edited_messages: bool = False) -> None:
        """Build the channel.

        Args:
            token: Telegram bot token; embedded in the Bot API base URL.
            http_client: Optional HTTP client override forwarded to
                ``http_post``.
            accept_edited_messages: When ``True``, ``edited_message`` updates
                are parsed and dispatched. Off by default to prevent
                replay-by-edit: a malicious sender can otherwise edit an
                old message repeatedly to keep re-triggering agent runs.
        """
        super().__init__()
        self.token = token
        self._http = http_client
        self._accept_edited = accept_edited_messages
        self._base = f"https://api.telegram.org/bot{token}"

    def _parse_inbound(self, headers, body):
        """Translate a Telegram update into ``ChannelMessage`` instances.

        Handles both ``message`` and (when ``accept_edited_messages``)
        ``edited_message`` updates. Captions of media messages are also
        used so a photo with an attached caption still produces text the
        agent can act on. Returns an empty list when there is no message
        envelope at all.

        Args:
            headers: HTTP headers (unused; provided for interface
                compatibility).
            body: Raw JSON webhook body.

        Returns:
            A single-element list with the parsed message, or empty when
            the update carries nothing relevant.
        """
        data = parse_json_body(body)
        message = data.get("message") or {}
        if not message and self._accept_edited:
            message = data.get("edited_message") or {}
        chat = message.get("chat") or {}
        sender = message.get("from") or {}
        if not message:
            return []
        text = message.get("text") or message.get("caption") or ""
        return [
            ChannelMessage(
                text=text,
                channel=self.name,
                channel_user_id=str(sender.get("id", "")),
                room_id=str(chat.get("id", "")),
                platform_message_id=str(message.get("message_id", "")),
                direction=MessageDirection.INBOUND,
                metadata={
                    "username": sender.get("username", ""),
                    "first_name": sender.get("first_name", ""),
                    "last_name": sender.get("last_name", ""),
                    "chat_type": chat.get("type", ""),
                    "chat_title": chat.get("title", ""),
                    "thread_id": str(message.get("message_thread_id", "")),
                },
            )
        ]

    def get_updates(self, *, offset: int | None = None, timeout: int = 30) -> dict[str, tp.Any]:
        """Long-poll the Telegram Bot API for new updates.

        Restricts the ``allowed_updates`` list to ``message`` (and
        ``edited_message`` when configured) so the bot does not receive
        callback queries or inline-mode traffic it cannot service.

        Args:
            offset: Last-seen update id + 1; tells Telegram which updates
                to skip. ``None`` returns the oldest unread batch.
            timeout: Long-poll timeout in seconds; Telegram will hold the
                connection that long before responding empty.

        Returns:
            The raw Telegram Bot API response.
        """
        allowed = ["message", "edited_message"] if self._accept_edited else ["message"]
        body: dict[str, tp.Any] = {"timeout": timeout, "allowed_updates": allowed}
        if offset is not None:
            body["offset"] = offset
        return http_post(f"{self._base}/getUpdates", json_body=body, http_client=self._http, timeout=timeout + 5)

    async def send_text(self, *, chat_id: str, text: str, reply_to: str | None = None) -> dict[str, tp.Any]:
        """Call ``sendMessage`` and return the raw Bot API response.

        Args:
            chat_id: Target chat id.
            text: Message body.
            reply_to: Optional message id to quote-reply to.
        """
        body: dict[str, tp.Any] = {"chat_id": chat_id, "text": text}
        if reply_to:
            body["reply_to_message_id"] = reply_to
        return http_post(f"{self._base}/sendMessage", json_body=body, http_client=self._http)

    async def edit_text(self, *, chat_id: str, message_id: str, text: str) -> dict[str, tp.Any]:
        """Replace the text of a previously sent message via ``editMessageText``.

        Used by ``TelegramAgentGateway`` to stream preview updates without
        spamming the chat with a new message per chunk.
        """
        body = {"chat_id": chat_id, "message_id": message_id, "text": text}
        return http_post(f"{self._base}/editMessageText", json_body=body, http_client=self._http)

    async def _send_outbound(self, message):
        """Deliver one outbound message via ``send_text``.

        Args:
            message: Outbound message. ``room_id`` (preferred) or
                ``channel_user_id`` selects the chat; ``text`` is the body;
                ``reply_to`` becomes ``reply_to_message_id`` when set.
        """
        await self.send_text(
            chat_id=message.room_id or message.channel_user_id,
            text=message.text,
            reply_to=message.reply_to,
        )

File: xerxes/src/channels/telegram.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import { timingSafeEqual } from 'node:crypto'

import { postJson, providerUrl, type ChannelFetch } from './http.js'
import { scanContextContent } from '../security/promptScanner.js'
import { createChannelMessage, MessageDirection, type ChannelMessage } from './types.js'
import {
  parseJsonBody,
  WebhookChannel,
  type WebhookHeaders,
} from './webhooks.js'

const TELEGRAM_API_BASE = 'https://api.telegram.org/'
const DEFAULT_MAX_PAYLOAD_BYTES = 256 * 1024
const PATH_REDACTION = /(?:\/Users\/[^\s'"]+|\/home\/[^\s'"]+|\/private\/[^\s'"]+|\/var\/[^\s'"]+|\/tmp\/[^\s'"]+|~\/\.xerxes[^\s'"]*)/g
const TRACEBACK_REDACTION = /Traceback \(most recent call last\):.*?(?=\n\n|$)/gs

type StringList = string | readonly string[] | ReadonlySet<string> | undefined

export interface TelegramChannelOptions {
  readonly acceptEditedMessages?: boolean
  /** Allowlist of Telegram user IDs. Used only when requireAllowedSender is enabled. */
  readonly allowedUserIds?: StringList
  /** Allowlist of Telegram usernames (with or without @). */
  readonly allowedUsernames?: StringList
  /** Override only for a Telegram-compatible API or tests. */
  readonly apiBaseUrl?: string
  /** Bot @username used for exact group-addressing checks. */
  readonly botUsername?: string
  readonly fetchImplementation?: ChannelFetch
  /** Per-channel hard cap enforced after the webhook body is read. */
  readonly maxPayloadBytes?: number
  /** Enforce a fail-closed Telegram sender allowlist. */
  readonly requireAllowedSender?: boolean | string
  readonly token: string
  /** Register this public callback URL when the channel starts in webhook mode. Requires webhookSecretToken. */
  readonly webhookUrl?: string
  /** Telegram's secret-token header; mandatory whenever webhookUrl is configured. */
  readonly webhookSecretToken?: string
}

export interface TelegramUpdatesOptions {
  readonly offset?: number
  readonly signal?: AbortSignal
  readonly timeout?: number
}

/** Telegram Bot API webhook adapter with optional long-poll helpers. */
export class TelegramChannel extends WebhookChannel {
  readonly name = 'telegram'
  readonly maxWebhookBodyBytes: number

  private readonly acceptEditedMessages: boolean
  private readonly allowedUserIds: ReadonlySet<string>
  private readonly allowedUsernames: ReadonlySet<string>
  private readonly apiBaseUrl: string
  private readonly botUsername: string
  private readonly fetchImplementation: ChannelFetch | undefined
  private readonly requireAllowedSender: boolean
  private readonly token: string
  private readonly webhookSecretToken: string
  private readonly webhookUrl: string

  constructor(options: TelegramChannelOptions) {
    super()
    if (!options.token) {
      throw new TypeError('Telegram bot token must not be empty')
    }
    this.token = options.token
    this.apiBaseUrl = options.apiBaseUrl ?? TELEGRAM_API_BASE
    this.fetchImplementation = options.fetchImplementation
    this.acceptEditedMessages = options.acceptEditedMessages ?? false
    this.allowedUserIds = identifierSet(options.allowedUserIds)
    this.allowedUsernames = usernameSet(options.allowedUsernames)
    this.botUsername = normalizedUsername(options.botUsername ?? '')
    this.maxWebhookBodyBytes = payloadLimit(options.maxPayloadBytes ?? DEFAULT_MAX_PAYLOAD_BYTES)
    this.requireAllowedSender = asBoolean(options.requireAllowedSender, false)
    this.webhookSecretToken = options.webhookSecretToken?.trim() ?? ''
    this.webhookUrl = options.webhookUrl?.trim() ?? ''
    if (this.webhookUrl && !this.webhookSecretToken) {
      throw new TypeError(
        'Telegram webhookUrl requires webhookSecretToken so forged updates cannot reach agent sessions',
      )
    }
  }

  override async start(onInbound: import('./base.js').InboundHandler): Promise<void> {
    await super.start(onInbound)
    if (!this.webhookUrl) return
    try {
      await this.setWebhook(this.webhookUrl)
    } catch (error) {
      await super.stop()
      throw error
    }
  }

  override async handleWebhook(
    headers: WebhookHeaders,
    body: Uint8Array,
  ): Promise<import('./webhooks.js').WebhookResponse> {
    if (body.byteLength > this.maxWebhookBodyBytes) {
      return { status: 413, body: 'payload too large' }
    }
    if (this.webhookSecretToken && !telegramSecretMatches(headers, this.webhookSecretToken)) {
      return { status: 401, body: 'unauthorized' }
    }
    return super.handleWebhook(headers, body)
  }

  /**
   * Ingest one update received through authenticated Bot API long polling.
   *
   * Polling is already authenticated by the bot token, so this path bypasses
   * the webhook-only secret-token check while retaining the shared parser and
   * inbound error containment.
   */
  async ingestPolledUpdate(body: Uint8Array): Promise<import('./webhooks.js').WebhookResponse> {
    if (body.byteLength > this.maxWebhookBodyBytes) {
      return { status: 413, body: 'payload too large' }
    }
    return super.handleWebhook({}, body)
  }

  /** Long-poll Telegram updates when a deployment does not expose a webhook. */
  async getUpdates(options: TelegramUpdatesOptions = {}): Promise<Readonly<Record<string, unknown>>> {
    const timeout = options.timeout ?? 30
    if (!Number.isInteger(timeout) || timeout < 0) {
      throw new RangeError('Telegram getUpdates timeout must be a non-negative integer')
    }
    const allowedUpdates = this.acceptEditedMessages
      ? ['message', 'edited_message']
      : ['message']
    const payload: Record<string, unknown> = {
      timeout,
      allowed_updates: allowedUpdates,
    }
    if (options.offset !== undefined) {
      payload.offset = options.offset
    }
    return asRecord(await this.request('getUpdates', payload, options.signal))
  }

  /** Send a Telegram text message and return Telegram's API envelope. */
  async sendText(
    chatId: string,
    text: string,
    replyTo?: string,
  ): Promise<Readonly<Record<string, unknown>>> {
    if (!chatId) {
      throw new TypeError('Telegram chat id must not be empty')
    }
    const payload: Record<string, unknown> = { chat_id: chatId, text: sanitizeTelegramOutbound(text) }
    if (replyTo) {
      payload.reply_to_message_id = replyTo
    }
    return asRecord(await this.request('sendMessage', payload))
  }

  /** Replace a previously sent Telegram message's text. */
  async editText(
    chatId: string,
    messageId: string,
    text: string,
  ): Promise<Readonly<Record<string, unknown>>> {
    if (!chatId || !messageId) {
      throw new TypeError('Telegram editText requires chatId and messageId')
    }
    return asRecord(await this.request('editMessageText', {
      chat_id: chatId,
      message_id: messageId,
      text: sanitizeTelegramOutbound(text),
    }))
  }

  /** Show Telegram's transient typing indicator for the target chat. */
  async sendTyping(chatId: string | undefined): Promise<void> {
    const target = chatId?.trim()
    if (!target) return
    await this.request('sendChatAction', { chat_id: target, action: 'typing' })
  }

  /** Register a public Telegram Bot API webhook with the configured secret token. */
  async setWebhook(webhookUrl: string): Promise<Readonly<Record<string, unknown>>> {
    const url = webhookUrl.trim()
    if (!url) throw new TypeError('Telegram webhook URL must not be empty')
    const payload: Record<string, unknown> = { url }
    if (this.webhookSecretToken) payload.secret_token = this.webhookSecretToken
    return asRecord(await this.request('setWebhook', payload))
  }

  /** Clear Telegram's registered webhook before switching to long polling. */
  async deleteWebhook(): Promise<Readonly<Record<string, unknown>>> {
    return asRecord(await this.request('deleteWebhook', {}))
  }

  protected parseInbound(
    _headers: WebhookHeaders,
    body: Uint8Array,
  ): readonly ChannelMessage[] {
    const update = parseJsonBody(body)
    const message = asRecord(update.message)
    const editedMessage = this.acceptEditedMessages ? asRecord(update.edited_message) : {}
    const envelope = Object.keys(message).length ? message : editedMessage
    if (!Object.keys(envelope).length) {
      return []
    }
    const sender = asRecord(envelope.from)
    const chat = asRecord(envelope.chat)
    const rawText = stringOrEmpty(envelope.text) || stringOrEmpty(envelope.caption)
    const userId = stringOrEmpty(sender.id)
    const username = stringOrEmpty(sender.username)
    const chatType = stringOrEmpty(chat.type)
    if (!this.senderAllowed(userId, username) || this.groupMessageIsNotAddressed(rawText, chatType)) {
      return []
    }
    return [createChannelMessage({
      channel: this.name,
      text: scanContextContent(rawText, 'telegram:inbound'),
      direction: MessageDirection.INBOUND,
      channelUserId: userId,
      roomId: stringOrEmpty(chat.id),
      platformMessageId: stringOrEmpty(envelope.message_id),
      metadata: {
        username,
        first_name: stringOrEmpty(sender.first_name),
        last_name: stringOrEmpty(sender.last_name),
        chat_type: chatType,
        chat_title: stringOrEmpty(chat.title),
        thread_id: stringOrEmpty(envelope.message_thread_id),
      },
    })]
  }

  protected async sendOutbound(message: ChannelMessage): Promise<void> {
    const chatId = message.roomId ?? message.channelUserId
    if (!chatId) {
      throw new TypeError('Telegram outbound messages require roomId or channelUserId')
    }
    await this.sendText(chatId, sanitizeTelegramOutbound(message.text), message.replyTo)
  }

  private senderAllowed(userId: string, username: string): boolean {
    if (!this.requireAllowedSender) return true
    if (!this.allowedUserIds.size && !this.allowedUsernames.size) return false
    return this.allowedUserIds.has(userId) || this.allowedUsernames.has(normalizedUsername(username))
  }

  private groupMessageIsNotAddressed(text: string, chatType: string): boolean {
    if (chatType.toLowerCase() !== 'group' && chatType.toLowerCase() !== 'supergroup') return false
    // Generic adapter configurations historically delivered every group message.
    // Exact-address filtering becomes active only once an operator supplies the
    // bot username needed to make the decision without guessing.
    if (!this.botUsername) return false
    const normalized = text.trim().toLowerCase()
    if (!normalized) return true
    const bot = this.botUsername
    if (bot && normalized.includes(`@${bot}`)) return false
    if (bot && (normalized.startsWith(`/${bot}`) || normalized.startsWith(`/xerxes@${bot}`))) return false
    return !(normalized === '/xerxes' || normalized.startsWith('/xerxes '))
  }

  private async request(method: string, body: unknown, signal: AbortSignal | undefined = undefined): Promise<unknown> {
    const result = await postJson(this.methodUrl(method), {
      body,
      ...(this.fetchImplementation ? { fetchImplementation: this.fetchImplementation } : {}),
      ...(signal ? { signal } : {}),
    })
    if (isRecord(result) && result.ok === false) {
      const description = typeof result.description === 'string'
        ? `: ${result.description}`
        : ''
      throw new Error(`Telegram API ${method} failed${description}`)
    }
    return result
  }

  private methodUrl(method: string): string {
    return providerUrl(this.apiBaseUrl, `bot${this.token}/${method}`)
  }
}

function asRecord(value: unknown): Readonly<Record<string, unknown>> {
  return isRecord(value) ? value : {}
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function stringOrEmpty(value: unknown): string {
  return value === undefined || value === null ? '' : String(value)
}

function identifierSet(value: StringList): ReadonlySet<string> {
  const items = typeof value === 'string' ? value.split(',') : value ? [...value] : []
  return new Set(items.map(item => String(item).trim()).filter(Boolean))
}

function usernameSet(value: StringList): ReadonlySet<string> {
  return new Set([...identifierSet(value)].map(normalizedUsername).filter(Boolean))
}

function normalizedUsername(value: string): string {
  return value.trim().replace(/^@/, '').toLowerCase()
}

function payloadLimit(value: number): number {
  if (!Number.isSafeInteger(value) || value < 1) {
    throw new RangeError('Telegram maxPayloadBytes must be a positive safe integer')
  }
  return value
}

function asBoolean(value: boolean | string | undefined, fallback: boolean): boolean {
  if (value === undefined) return fallback
  if (typeof value === 'boolean') return value
  return !new Set(['0', 'false', 'no', 'off', '']).has(value.trim().toLowerCase())
}

function telegramSecretMatches(headers: WebhookHeaders, expected: string): boolean {
  const actual = Object.entries(headers)
    .find(([name]) => name.toLowerCase() === 'x-telegram-bot-api-secret-token')?.[1] ?? ''
  const actualBytes = Buffer.from(actual)
  const expectedBytes = Buffer.from(expected)
  return actualBytes.byteLength === expectedBytes.byteLength && timingSafeEqual(actualBytes, expectedBytes)
}

function sanitizeTelegramOutbound(text: string): string {
  if (!text) return text
  return text
    .replace(TRACEBACK_REDACTION, '[traceback redacted]')
    .replace(PATH_REDACTION, '[path redacted]')
}
